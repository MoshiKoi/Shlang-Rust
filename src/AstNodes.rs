
use std::collections::*;
use std::rc::Rc;
#[derive(Clone,Debug,PartialEq)]
pub enum Control {
    Return(Node),
    Break
}
#[derive(Clone,Debug,PartialEq)]
pub enum Value{
    NoneType,
    Control(Control),
    Num(f32),
    Bool(bool),
    Str(String),
    Function(Function),
    BuiltinFunc(BuiltinFunc)
}
#[derive(Clone,Debug,PartialEq)]
pub struct Function {
    block:Box<Block>,
    args:Vec<String>
}
#[derive(Clone,Debug,PartialEq)]
pub struct BuiltinFunc{
    funcPointer:fn(Vec<Value>) -> Value,
    argSize:i16
}
type ValueStream = Vec<Value>;

#[derive(Clone,Debug,PartialEq)]
pub enum Node {
    Value(Box<Value>),
    Block(Block),
    BinaryNode(BinaryNode),
    UnaryNode(UnaryNode),
    Declaration(Declaration),
    Assignment(Assignment),
    Variable(Variable),
    Call(Call),
    BranchNode(BranchNode),
    LoopNode(LoopNode),
    WhileNode(WhileNode),    
}

impl Node {
    pub fn declaration(var_name:String,value:Node)-> Self {
        return Node::Declaration(Declaration{
            var_name,
            value:Box::new(value)
        });
    }
    pub fn block(body:NodeStream)-> Self {
        return Node::Block(Block { body});
    }
}
impl From<Function> for Value {
    fn from(x: Function) -> Self {
      Value::Function(x)
    }
}
impl From<Value> for Node {
    fn from(x: Value) -> Self {
      Node::Value(Box::new(x))
    }
}
macro_rules! nodes_from {
    ($($name:ident)*) => {
        $(
            impl ::core::convert::From<$name> for Node {
                fn from(node: $name) -> Self {
                    Self::$name(node)
                }
            }
        )*
    }
}
nodes_from! { UnaryNode BinaryNode Call Variable Assignment Declaration BranchNode WhileNode LoopNode Block}
pub type NodeStream = Vec<Node>;
pub type NodeRef = Box<Node>;
#[derive(Clone,Debug,PartialEq)]
pub struct Block {
    pub body:NodeStream
}
pub type BlockRef = Box<Block>;
#[derive(Clone,Debug,PartialEq)]
pub enum BinaryOp {
    ADD,
    SUBTRACT,
    DIVIDE,
    MULTIPLY,
    MODULO,
    AND,
    OR,
    ISEQUAL,
    ISDIFERENT,
    GREATER,
    LESSER,
    GREATER_EQUAL,
    LESSER_EQUAL,
}
#[derive(Clone,Debug,PartialEq)]
pub struct BinaryNode {
    pub kind:BinaryOp,
    pub left:NodeRef,
    pub right:NodeRef
}
#[derive(Clone,Debug,PartialEq)]
pub enum UnaryOp{
    NEGATIVE,
	POSITIVE,
	NOT,
}
#[derive(Clone,Debug,PartialEq)]
pub struct UnaryNode {
    pub kind:UnaryOp,
    pub object:NodeRef,
}
#[derive(Clone,Debug,PartialEq)]
pub struct Declaration{
    pub var_name:String,
    pub value:NodeRef
}
impl Declaration {
    fn new(var_name:String,v: impl Into<NodeRef>) -> Self { 
        Self{var_name, value:v.into()} 
    }
}
#[derive(Clone,Debug,PartialEq)]
pub struct Assignment{
    pub var_name:String,
    pub value:NodeRef
}
impl Assignment {
    fn new(var_name:String,v: impl Into<NodeRef>) -> Self { 
        Self{var_name, value:v.into()} 
    }
}
#[derive(Clone,Debug,PartialEq)]
pub struct Variable{
    pub name:String
}
#[derive(Clone,Debug,PartialEq)]
pub struct Call{
    pub callee:Variable,
    pub args:Box<NodeStream>,
}
#[derive(Clone,Debug,PartialEq)]
pub struct BranchNode{
    pub condition:NodeRef,
    pub if_block:BlockRef,
    pub else_block:Option<BlockRef>
}
#[derive(Clone,Debug,PartialEq)]
pub struct LoopNode{
    pub proc:BlockRef,
}
#[derive(Clone,Debug,PartialEq)]
pub struct WhileNode{
    pub condition:NodeRef,
    pub proc:BlockRef,
}
#[derive(Clone,Debug,PartialEq)]
pub struct Scope {
    pub parent:Option<Box<Scope>>,
    pub var_map:HashMap<String,Value>
}
impl Scope {
pub fn get_var(&self,var_name:String) -> Option<Value>{
    if let Some(var) = self.var_map.get(&var_name) {
      return Some(var.clone());
    }    
    if let Some(parent) = &self.parent {
      return parent.get_var(var_name);
    }
    
    None
}
pub fn define(&mut self,var_name:String,val:Value){
    self.var_map.insert(var_name, val);
}
fn assign(&mut self, var_name: String, value: Value) -> Option<Value>{
    if let Some(var) = self.var_map.get_mut(&var_name) {
        *var = value;
        return Some(var.clone());
    }    
    if let Some(parent) = &mut self.parent {
        return parent.assign(var_name,value);
    }
    None
}
}