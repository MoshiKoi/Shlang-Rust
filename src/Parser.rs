use crate::Token::*;
use crate::AstNodes::*;
struct Parser {
    stream: Vec<Token>,
    current: usize,
}

impl Parser {
    fn peek_next(&mut self) -> Token {
        self.stream[self.current + 1]
    }
    fn peek(&mut self) -> Token {
        self.stream[self.current]
    }
    fn expect(&mut self, what: Token) -> bool {
        self.peek_next() == what
    }
    fn next(&mut self) -> Token {
        self.current += 1;
        self.stream[self.current]
    }
    
    pub fn parse(&self) -> Node {
        
    }
    fn new(stream: Vec<Token>)->Self{
        Parser{stream:stream,current:0}
    }
}
