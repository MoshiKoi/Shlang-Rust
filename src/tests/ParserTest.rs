use crate::AstNodes;
use crate::Parser;
use std::io;
use std::io::Write;
use std::*;
use AstNodes::*;
use insta::*;
fn parse_text(source: &str) -> Node {
    let mut parser = Parser::new(source);
    return parser.batch_parse();
}
macro_rules! create_function {
    ($($name:ident : $input:expr)*) => {
        $(
        #[test]
        fn $name () {
            // Perform your desired operation using the input parameter
            assert_debug_snapshot!(parse_text($input));
        }
        )*
    };
}
create_function!(
declare_and_assign_expression:"var a = 1+2+b;" 
single_var_decl:"var a;"
multi_var_decl:"var a; var a;"
var_declare_and_assign_value:"var a = 2;"
parse_nested_paren:"var a = ((1+2)+(1+2));"
parse_paren:"var a = (1+2+b);"
expression_operators:"var a = 1*2+2-2/2%5 and 10 or 10!=1==2<4<=3>0>=3;"
unary_operator:"var a = !true;"
multiple_unary:"var a = !true + !true;"
unary_operator_paren:"var a = !(0);"
unary_operators:"var a = !0 +(+0)+(-0);"
);
