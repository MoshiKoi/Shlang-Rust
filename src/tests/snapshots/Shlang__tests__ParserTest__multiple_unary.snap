---
source: src/tests/ParserTest.rs
expression: "parse_text(\"var a = !true + !true;\")"
---
Block(
    Block {
        body: [
            Declaration(
                Declaration {
                    var_name: "a",
                    value: BinaryNode(
                        BinaryNode {
                            kind: ADD,
                            left: UnaryNode(
                                UnaryNode {
                                    kind: NOT,
                                    object: Value(
                                        Bool(
                                            true,
                                        ),
                                    ),
                                },
                            ),
                            right: UnaryNode(
                                UnaryNode {
                                    kind: NOT,
                                    object: Value(
                                        Bool(
                                            true,
                                        ),
                                    ),
                                },
                            ),
                        },
                    ),
                },
            ),
        ],
    },
)
